src/browser/fetch_network.js

```diff
@@ -113,13 +115,41 @@ async function on_data_http(data)
             const header = this.net.parse_http_header(headers[i]);
             if(!header) {
                 console.warn('The request contains an invalid header: "%s"', headers[i]);
-                this.write(new TextEncoder().encode("HTTP/1.1 400 Bad Request\r\nContent-Length: 0"));
+                const body = new TextEncoder().encode(`Invalid header in request: ${headers[i]}`);
+                const resp_headers = new Headers({
+                    "content-type": "text/plain",
+                    "content-length": body.length.toString(10),
+                    "connection": "close"
+                });
+                this.writev([new TextEncoder().encode(this.net.form_response_head(400, "Bad Request", resp_headers)), body]);
+                this.close();
                 return;
             }
             if( header.key.toLowerCase() === "host" ) target.host = header.value;
             else req_headers.append(header.key, header.value);
         }
 
+        if(this.net.local_http && !this.net.cors_proxy && (/\d+\.v86local\.http/).test(target.hostname)) {
+            dbg_log("Request to localhost: " + target.href, LOG_FETCH);
+            const localport = parseInt(target.hostname.split(".")[0], 10);
+            if(!isNaN(localport) && localport > 0 && localport < 65536) {
+                target.protocol = "http:";
+                target.hostname = "localhost";
+                target.port = localport.toString(10);
+            } else {
+                console.warn('Unknown port for localhost: "%s"', target.href);
+                const body = new TextEncoder().encode(`Unknown port for localhost: ${target.href}`);
+                const resp_headers = new Headers({
+                    "content-type": "text/plain",
+                    "content-length": body.length.toString(10),
+                    "connection": "close"
+                });
+                this.writev([new TextEncoder().encode(this.net.form_response_head(400, "Bad Request", resp_headers)), body]);
+                this.close();
```

Comment:
Could be moved into a function (also used in some other places):


```javascript
                const body = new TextEncoder().encode(`Unknown port for localhost: ${target.href}`);
                const resp_headers = new Headers({
                    "content-type": "text/plain",
                    "content-length": body.length.toString(10),
                    "connection": "close"
                });
                this.respond_and_close(400, "Bad request", resp_headers, body);
```

Reply:
I would make a method to respond with a message/error:
```js
this.respond_message_and_close(400, "Bad request", `Unknown port for localhost: ${target.href}`);
```